src/NuGet.Core/NuGet.ProjectModel/CacheFileFormat.cs

```diff
@@ -5,49 +5,172 @@
 using System.Collections.Generic;
 using System.Globalization;
 using System.IO;
-using Newtonsoft.Json;
-using Newtonsoft.Json.Linq;
+using System.Linq;
+using System.Text.Encodings.Web;
+using System.Text.Json;
+using System.Text.Json.Serialization;
 using NuGet.Common;
 
 namespace NuGet.ProjectModel
 {
     public static class CacheFileFormat
     {
-        private const string VersionProperty = "version";
-        private const string DGSpecHashProperty = "dgSpecHash";
-        private const string SuccessProperty = "success";
-        private const string ExpectedPackageFilesProperty = "expectedPackageFiles";
-        private const string ProjectFilePathProperty = "projectFilePath";
-
-        public static CacheFile Read(Stream stream, ILogger log, string path)
+        private static JsonSerializerOptions SerializerOptions = new JsonSerializerOptions
         {
-            if (stream == null) throw new ArgumentNullException(nameof(stream));
-            if (log == null) throw new ArgumentNullException(nameof(log));
-            if (path == null) throw new ArgumentNullException(nameof(path));
+            WriteIndented = true,
+            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+            Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
+            DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull,
+            Converters = { new AssetsLogMessageConverter() }
+        };
+
+        /// <summary>
+        /// since Log messages property in CacheFile is an interface type, we have the following custom converter to deserialize the IAssetsLogMessage objects.
+        /// </summary>
+        private class AssetsLogMessageConverter : JsonConverter<IAssetsLogMessage>
```

Comment:
Part of the idea of calling the LockFileFormat implementation was to avoid maintaining 2 different implementations of the assets file log message reading. 

Is there anything in that implementation we can reuse given that there's an STJ implementation?

Reply:
You are right. LockFileFormat has implementations for reading and writing IAssetsLogMessage. However, as you mentioned, it is [currently implemented](https://github.com/NuGet/NuGet.Client/blob/764d370df4d3c01c24340f57e6389058bce24ddf/src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileFormat.cs#L496) using NSJ. I don't think we could reuse it currently for a STJ implementation. I think, in the future, when we migrate LockFileFormat, this converter could be used by every method trying to read and write asset file log messages.

Maybe if we modify ReadLogMessage in LockFileFormat to take in a delegate that allows it to read its input, we could be able to use it to read log messages from both` NSJ JObject` and `STJ JsonElement`. The new ReadLogMessage, could have a definition that looks like 
* delegate `public delegate T JsonPropertyReader<T>(object json, string property)`
* `private static IAssetsLogMessage ReadLogMessage(
    object json, 
    string projectPath, 
    JsonPropertyReader<string> readStringProperty, 
    JsonPropertyReader<int> readIntProperty)`

My fear is we might introduce some allocation overhead from the delegates.

Reply:
The lock file reading is implementing with STJ too: https://github.com/NuGet/NuGet.Client/commit/ef5bce328208c10d00beaf00a96a189af30cfd0b. 

It's what's used by default. 
There seems to be a converter implemented there: https://github.com/NuGet/NuGet.Client/blob/f7ae70859a8aa9c3f3fda6a3170bbfd437e47272/src/NuGet.Core/NuGet.ProjectModel/Utf8JsonReaderExtensions.cs#L19. 

I was thinking we can use that: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.ProjectModel/LockFile/Utf8JsonStreamIAssetsLogMessageConverter.cs.

Reply:
unfortunately, not really. STJ's `Utf8JsonReader` doesn't have a `Read()` or `MoveNext()` method that can read from a stream. While STJ can internally do a reading stream with by-reflection (or source generator) serializer, if you implement their `JsonConverter`, it will buffer the entire object in memory. Since we'd need a converter for the root object, this means that STJ would need `JsonSerializer.Deserialize<AssetsFile>(stream)` to buffer the entire assets file in memory, and since that can be multiple megabytes, we don't do that.

STJ's docs recommend you implement your own streaming JsonReader, which is what we did for the assets file, and hence `IUtf8JsonStreamReaderConverter<T>` is a custom NuGet thing: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReaderConverter.cs#L9

This means it can't (easily?) be reused with STJ's `JsonConverter<T>` or `JsonSerializer.Deserialize<T>()`.

Reply:
The largest part of the implementation for CacheFileFormat is the log messages. 
I know it's not the recommended way, but I'm wondering if we're better off just following that pattern with CacheFileFormat. 

I'm just trying to further minimize code duplication with serialization. We'll need 2 to implement STJ, I'd hate to have 3.

Reply:
If we're not incorporating my feedback to use the existing converter, let's at least add detailed tests about the converter. 

Replaying the restore messages correctly is really important tot he experience given how often a restore no-ops.

Reply:
Sorry for not giving this feedback earlier. For some reason, I only just thought of it now 😕 

The implementation of the `Read` method can be changed to:

```cs
var converter = (JsonConverter<AssetsLogMessage>)options.GetConverter(typeof(AssetsLogMessage>));
return converter.Read(ref reader, typeof(AssetsLogMessage), options);
```

It might require annotating AssetsLogMessage to be compatible with System.Text.Json, similar to how CacheFile was annotated.  But then we can make the `Read` method only 2 or 3 lines long, not 70, and similarly the Write method can remove nearly 50 lines.


Also, I'm kind of wondering if we should break the public API and change `CacheFile`'s property to `IList<AssetsLogMessage>`, or `IReadOnlyList<AssetsLogMessage>` (or `ImmutableArray<AssetsLogMessage>`), instead of `IList<IAssetsLogMessage>`. 

Considering that `AssetsLogFile` is a DTO, there's no business logic in it, I honestly don't understand the point of `IAssetsLogMessage`, and I'm a fairly strong believer in the SOLID design principals and the port-adapter pattern.  Anyway, CacheFile could use the concrete class directly, and then we wouldn't need a 10 line converter. 
What do you think @nkolev92?

Reply:
I'm guessing you're asking me with thoughts on breaking the public API for CacheFile?
I added CacheFile, and if I had known what I know now, I would've kept that in NuGet.Commands instead of NuGet.ProjectModel, and nothing would've been public. 
I don't expect anyone to be using this, so as long as we document the change, i'm ok with breaking the API if you 2 agree we want that.

Reply:
I like the idea of `IList<AssetsLogMessage>`. I will make create another PR updating our API and then  I will update this PR

Reply:
https://github.com/NuGet/NuGet.Client/pull/6132

Reply:
@Nigusu-Allehu I'll rely on @zivkan here, but I don't think you need that change. 

You can just assume the concrete class in that converter and you should be good.